src/reducers-actions/trades.ts

```diff
@@ -29,30 +39,99 @@ export const updateLastCheckedBlock = (lastCheckedBlock: number): UpdateBlockAct
+
+  // Assumptions:
+  // 1. There can be more than one trade per batch for a given order (even if there are no reverts)
+  //    This case is not likely given our current solvers, but it's not forbidden by the contract
+  // 2. There will never be more reverts than trades for a given batch/order pair
+  // 3. Every revert matches 1 trade
+  // 4. Reverts match Trades by order or appearance (first Revert matches first Trade and so on)
+
+  Array.from(revertsByRevertKey.keys()).forEach(revertKey => {
+    const reverts = revertsByRevertKey.get(revertKey) as TradeReversion[]
+    const trades = tradesByRevertKey.get(revertKey)
+
+    // Given the assumptions above, the possibilities are:
+    // 1. # trades > # reverts
+    // 2. # trades == # reverts
+    // Not possible:
+    // 3. # trades < # reverts
+    // 4. There are reverts but there are no trades for given revertKey <- should NOT ever happen
+    // 5. There are trades but no reverts <- wouldn't be in this loop in that case
+    if (trades) {
+      const remainingTrades = trades.length - reverts.length
+      if (remainingTrades < 1) {
+        // Case 2 and 3, all trades reverted
+        tradesByRevertKey.delete(revertKey)
+        logDebug(`[reducers:trade][${revertKey}] All ${trades.length} trade(s) reverted by ${reverts.length} revert(s)`)
+      } else {
+        // Case 1. One or more trades left, pick from the end
+        const filteredTrades = trades.slice(trades.length - remainingTrades)
+        tradesByRevertKey.set(revertKey, filteredTrades)
+        logDebug(`[reducers:trade][${revertKey}] Reverted ${reverts.length} trade(s), ${filteredTrades.length} left`)
+      }
+    }
+    // TODO: this effectively makes the reverts global state to always be empty.
+    //    Should we care? Is there a use case where a revert will happen before a trade?
+    // Reverts have been "used", remove them
+    revertsByRevertKey.delete(revertKey)
+  })
+
+  return [flattenGroup(tradesByRevertKey), flattenGroup(revertsByRevertKey)]
+}
 
 export const reducer = (state: TradesState, action: ReducerActionType): TradesState => {
   switch (action.type) {
     case 'APPEND_TRADES': {
-      const { trades: currTrades } = state
-      const { trades: newTrades, lastCheckedBlock } = action.payload
-
-      const seenTradeIdsSet = new Set<string>()
-
-      // Append new list to existing list of trades
-      const trades = currTrades.concat(newTrades).reduce<Trade[]>((acc, trade) => {
-        // Removes possible duplicates
-        if (!seenTradeIdsSet.has(trade.id)) {
-          acc.push(trade)
-          seenTradeIdsSet.add(trade.id)
-        }
-        return acc
-      }, [])
-
-      return { trades, lastCheckedBlock }
+      const { trades: currTrades, reverts: currReverts } = state
+      const { trades: newTrades, reverts: newReverts, lastCheckedBlock } = action.payload
+
+      const [trades, reverts] = applyRevertsToTrades(currTrades.concat(newTrades), currReverts.concat(newReverts))
```

Comment:
This is not incremental, right?
I mean, you apply, and repply for past batches every time you have new trades

wouldn't it make sense to just have a global state with trades, this trades have a revertDate (in case they were reverted), and then you only need to use the reducer to appply the new information:
- new trades: You add them to the global state
- new reverts: You match them in order to the trades of that user/batch/orderId using blockNumber and the index of the event to sort them and decide which one is next

This way:
* You keep all the information as a single entity TRADES, that you know they can be reverted or not (checking the revertDate field)
* You process things incrementally, no need to reprocess past batches
* The hook can return this trades, so the view can decide to filter out and remove the reverted

Alternativelly, the hook could return the not reverted ones only, if we think we won't ever need the other ones. Maybe is good

Reply:
I see. Will give the idea of not deleting and instead setting a `revertDate` a try.

Reply:
Done: d3c154ab0e31df91097ae1f084bf0ba557438ff8

To illustrate, I added an icon for reverted trades (will remove this before merging)
![screenshot_2020-06-08_14-09-35](https://user-images.githubusercontent.com/43217/84080834-b168aa00-a991-11ea-8a2a-bf7fee270d49.png)

Reply:
Well, part of this comment was also originated by the line:

```js
const [trades, reverts] = applyRevertsToTrades(currTrades.concat(newTrades), currReverts.concat(newReverts))
```

It seams that on every `APPEND_TRADES`, so on every new trade, it looks like we reprocess all trades.

Issues with that:

* It's imposible to revert from the last batch. So when we change batch trades are final. An user with trades expanding months could reprocess only the last batch
* Maybe is just keeping track of the last batch you know, and the list of unreverted trades on it. Then you map 1<>1 in order for the reverts and remove them from the list (they never need to be processed again either)
* The double loop is a bit complex, iterating with 2 indexes. Maybe if you just keep the `unreverted` you can use a reduce like this:

```js
// Recalculate unreverted trades for the current batch
unreverted = reverts.reduce(youReduceLogic, new Set(unreverted))
```

Reply:
Also, for first processing, I understand we need to process old batches, so the logic is a bit more complex.

I tried to write this code in case it gives you any ideas, not for using it right away, cause I know it won't work (I think I ignored for example that the matching should consider the orderIds), so take it just as an idea on incremental updating.

```js
// Group new trades and reverts by batch
// Note how I make sure I also consider my previous unrevertedTrades  from my comment above  ☝️
const tradesByBatch: BatchInfo<Trade> = trades.concat(unrevertedTrades).reduce(groupByBatchId, [])
const revertsByBatch: BatchInfo< EventWithBlockInfo> = reverts.reduce(groupByBatchId, [])

// not sure if we can assume order, otherwise
tradesByBatch = tradesByBatch.sort(sortByBatchId)


// Mark some trades as reverted, and save the unreverted of the batch
tradesByBatch.tradesByBatch(batch => {
   let reverts = revertsByBatch.find(batchRevert => batchRevert.batchId === batch.batchId) // you can use a map, if u find u want to be more performant
  
     // Make sure revers and trades are in order
    const trades = batch.items.sort(sortByBlockNumberAndEventIndex)
    reverts = reverts ? reverts.sort(sortByBlockNumberAndEventIndex) : []

    // Get the unreverted trades
    let unreverted = trades.filter(notReverted)

    // Recalculate unreverted trades for the current batch
    unreverted = reverts.reduce(yourReduceLogic, new Set(unreverted))
    batch.unreverted = unreverted
})

// Keep track always of the latests unreverted trades (only for the current batch), so when we get new trades and reverts we can consider them for the incremental update
const latestBatch = tradesByBatch[tradesByBatch.length -1]
unrevertedTrades = latestBatch >= currentBatch ? latestBatch.unreverted : []
```

Reply:
I get your concern, and I can address parts, but not all of it.

Even if I chose to (re)process only the last N batches (N = 1, 2, etc) on insertion, I'll have to go over the current list and find trades of given batches.
That's the same cost already "paid" for grouping trades (by batch|order).
I could go back only X trades looking for N batches to shorten the search range, to improve this a bit.

One thing that I don't think I can do is to ignore already reverted trades.
That's because for a given Revert, I can only tell it has been consumed when I find a trade with a matching `revertId`. Otherwise I risk (in the worst case) reverting all trades (of a batch|order) with a single Revert event.

One solution (to not processing all trades on every updates) would be to store in the global state the map with groups of trades for faster insertion, at the cost of flattening the map on read.

---

Alternatively, (re-read multiple times your points) trying to interpret your suggestion again.

1. On first load, match reverts to trades. 
1. Store on global state all trades from current + last batch as `pendingTrades`.
1. On every new group of trades/reverts:
    1. Add `pendingTrades` to new `trades`
    1. Apply `reverts` to them.
    1. Update global trades list with the result
    1. Repeat step `2`

Is this more inline with what you are thinking?
I'll give this a try.

---

Last point: double index loop is necessary! 
The alternative would be a less efficient nested loop, where for each trade I'll loop through (potentially all) reverts. Even if I remove consumed reverts from this list, it's not as good.

Reply:
I think we discuss it in the chat. let me know if you want to discuss anything further

> Last point: double index loop is necessary!
😂

Reply:
Hopefully now I got what you meant. I addressed somewhat as I described on **Part 2** above.

And **Part 3** remains!